### PR TITLE
[dev-11.0] Make it possible to hide site email address

### PR DESCRIPTION
**Pull request by EreMaijala**
This allows one to hide site email address when no support requests are wanted via email.

### CODE REVIEW

File: module/VuFind/src/VuFind/Config/Feature/EmailSettingsTrait.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+     * @return string
+     */
+    protected function getEmailSenderAddress(array|Config $config, ?string $userEmail = null): string
+    {
+        if ($config instanceof Config) {
+            $config = $config->toArray();
+        }
+        return (null !== $userEmail && ($config['Mail']['user_email_in_from'] ?? false))
+            ? $userEmail
+            : $config['Mail']['default_from'] ?? $config['Site']['email'];
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Config/Feature/EmailSettingsTrait.php]: It seems like it's now possible to have a situation where none of these settings are set, since the new comment on `$config['Site']['email']` appears to suggest that the setting is now optional. Should we throw an exception if we can't get a string of any sort? Does the comment need to be clarified in some way?

File: module/VuFind/src/VuFind/Config/Feature/EmailSettingsTrait.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+/**
+ * Trait providing email settings
+ *
+ * @category VuFind
+ * @package  Config
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+trait EmailSettingsTrait
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Config/Feature/EmailSettingsTrait.php:45]: I wonder if this should be EmailSenderSettingsTrait (or something else more specific) since other user-oriented email settings are handled by \VuFind\Config\AccountCapabilities. Or maybe we should just add a comment noting that for the benefit of anyone who ends up here looking for that code.

File: module/VuFind/tests/integration-tests/src/VuFindTest/Mink/EmailAuthenticationTest.php [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        $this->findCssAndSetValue($page, '.modal-body [name="username"]', 'catuser@vufind.org');
+        $this->clickCss($page, '.modal-body .btn.btn-primary');
+        $this->assertEquals(
+            'We have sent a login link to your email address. It may take a few moments for the link to arrive.'
+            . " If you don't receive the link shortly, please check also your spam filter.",
+            $this->findCssAndGetText($page, '.alert-success')
+        );
+
+        // Extract the link from the provided message:
+        $email = file_get_contents($this->getEmailLogPath());
```
Comment (demiankatz) [inline module/VuFind/tests/integration-tests/src/VuFindTest/Mink/EmailAuthenticationTest.php]: I assume you'll want to update this to match the improved behavior in #3975. Would it make sense to pull this test out to a separate PR so we don't lose track, since nothing else here really depends on it? Or should we just add a TODO checkbox somewhere?

File: module/VuFind/tests/integration-tests/src/VuFindTest/Mink/EmailAuthenticationTest.php [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    {
+        static::failIfDataExists();
+    }
+
+    /**
+     * Test the email authentication process.
+     *
+     * @return void
+     */
+    public function testEmailAuthentication(): void
```
Comment (demiankatz) [inline module/VuFind/tests/integration-tests/src/VuFindTest/Mink/EmailAuthenticationTest.php:64]: Should this be testILSEmailAuthentication, since there's also a separate ability to use Email authentication directly which might merit a separate testStandaloneEmailAuthentication test case at some point?

File: themes/bootstrap3/templates/error/contact.phtml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<?=$this->transEsc('Please contact the Library Reference Department for assistance')?>
```
Comment (demiankatz) [inline themes/bootstrap3/templates/error/contact.phtml]: While we're refactoring this anyway, how would you feel about making this two separate `<p>` tags inside this template, instead of using a `<br>` here and wrapping the render in `<p>`? That seems like it might be cleaner markup, though I won't argue further if you disagree. :-)

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @EreMaijala, this is a good idea! Just a few minor thoughts and suggestions.

File: module/VuFind/tests/integration-tests/src/VuFindTest/Mink/EmailAuthenticationTest.php (hunk 3 above)
Comment (demiankatz) [reply module/VuFind/tests/integration-tests/src/VuFindTest/Mink/EmailAuthenticationTest.php]: ...and now that #3975 is merged, you can just address it here if you wish. :-)

File: module/VuFind/src/VuFind/Config/Feature/EmailSettingsTrait.php (hunk 1 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Config/Feature/EmailSettingsTrait.php]: I updated the comment for Site email and added a check and exception to the trait.

File: module/VuFind/src/VuFind/Config/Feature/EmailSettingsTrait.php (hunk 2 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Config/Feature/EmailSettingsTrait.php:45]: I added a comment. I don't want to make this too specific to make it possible to add more (non-sender) settings later on if needed.

File: module/VuFind/tests/integration-tests/src/VuFindTest/Mink/EmailAuthenticationTest.php (hunk 3 above)
Comment (EreMaijala) [reply module/VuFind/tests/integration-tests/src/VuFindTest/Mink/EmailAuthenticationTest.php]: Done!

File: module/VuFind/tests/integration-tests/src/VuFindTest/Mink/EmailAuthenticationTest.php (hunk 4 above)
Comment (EreMaijala) [reply module/VuFind/tests/integration-tests/src/VuFindTest/Mink/EmailAuthenticationTest.php:64]: Changed, and added a test for Database+Email authentication as well.

File: themes/bootstrap3/templates/error/contact.phtml (hunk 5 above)
Comment (EreMaijala) [reply themes/bootstrap3/templates/error/contact.phtml]: Good idea, now done.

File: themes/bootstrap5/templates/error/index.phtml [hunk 6]
```diff
@@ -7,12 +7,7 @@
 <div class="alert alert-danger">
   <p><?=$this->transEsc('An error has occurred')?></p>
   <p><?=$this->transEsc($this->message)?></p>
-  <p>
-    <?=$this->transEsc('Please contact the Library Reference Department for assistance')?>
-    <br>
-    <?php $supportEmail = $this->escapeHtmlAttr($this->systemEmail()); ?>
-    <a href="mailto:<?=$supportEmail?>"><?=$supportEmail?></a>
-  </p>
+  <p><?=$this->render('error/contact.phtml')?></p>
```
Comment (demiankatz) [inline themes/bootstrap5/templates/error/index.phtml]: Now that error/contact.phtml contains complete `<p>` tags of its own, I think you can remove the wrapping paragraph in all the templates that include it.

File: module/VuFind/tests/integration-tests/src/VuFindTest/Mink/EmailAuthenticationTest.php [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+        $session = $this->getMinkSession();
+        $session->visit($this->getVuFindUrl());
+        $page = $session->getPage();
+
+        // Request login:
+        $this->clickCss($page, '#loginOptions a');
+        $this->findCssAndSetValue($page, '.modal-body #login_Email_username', 'username1@foo.bar');
+        $this->clickCss($page, '.modal-body .btn.btn-primary', null, 1);
+        $this->assertEquals(
+            'We have sent a login link to your email address. It may take a few moments for the link to arrive.'
```
Comment (demiankatz) [inline module/VuFind/tests/integration-tests/src/VuFindTest/Mink/EmailAuthenticationTest.php:132]: Would it be better if we informed the user that they provided an invalid email address, rather than acting like things were successful? Or am I misunderstanding the context here?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @EreMaijala, just a couple more minor questions/comments.

File: themes/bootstrap5/templates/error/index.phtml (hunk 6 above)
Comment (EreMaijala) [reply themes/bootstrap5/templates/error/index.phtml]: Right, of course! Now fixed.

File: module/VuFind/tests/integration-tests/src/VuFindTest/Mink/EmailAuthenticationTest.php (hunk 7 above)
Comment (EreMaijala) [reply module/VuFind/tests/integration-tests/src/VuFindTest/Mink/EmailAuthenticationTest.php:132]: No, we don't want to disclose any information to a potential perpetrator. Perhaps the message could be formatted in a different way, but the main point is to not tell if the email address was found or not.

Comment (demiankatz) [reply module/VuFind/tests/integration-tests/src/VuFindTest/Mink/EmailAuthenticationTest.php:132]: Maybe I misunderstood -- I of course agree that we don't want to disclose whether or not a valid email exists in the database, but I thought this case was to account for the user typing something that isn't an email address at all, in which case a separate message would not disclose anything sensitive and might help somebody who made a typo.

Comment (EreMaijala) [reply module/VuFind/tests/integration-tests/src/VuFindTest/Mink/EmailAuthenticationTest.php:132]: Right, of course. In that regard we should do a better job. Perhaps making the input field `type="email"` would make sense too. But I think it's out of scope here.

**Review by demiankatz — APPROVED** [review]
Thanks, @EreMaijala!
